### PR TITLE
Solving "expected a header" error 

### PR DESCRIPTION
**Pull request by vladyslav-dubina**
For testing we can use tests: `issue1897-bmv2.p4` , `invalid-hdr-warnings6.p4` , `issue561-bmv2.p4` and `invalid-hdr-warnings5.p4` with `--loopsUnroll` argument.

### CODE REVIEW

**Comment by vladyslav-dubina** [comment]
The falls are caused by very strange behavior in https://github.com/vladyslav-dubina/p4c/blob/3d5cd9d0ec8cfef5b3881503d35e33896723c5e9/midend/interpreter.cpp#L414 . `to<SymbolicBool>()`  gives the result `error: Null result` but `is<SymbolicBool>()` gives the result `true` . 
I can not understand what caused this

**Comment by fruffy** [comment]
I assume it is because of this line
https://github.com/vladyslav-dubina/p4c/blob/3d5cd9d0ec8cfef5b3881503d35e33896723c5e9/midend/interpreter.h#L400
Can you check whether validity is initialized correctly and add appropriate assertions?

**Comment by vladyslav-dubina** [comment]
In test `psa-fwd-bmv2.p4` validity is initialized to false

**Comment by fruffy** [comment]
Well, I would just temporarily check whether `valid` and/or its clone is a nullptr at this stage using `CHECK_NULL`. This seems like unusual behavior and I do not think your fix addresses the core of the issue.  

For example, what does it even mean that you are calling setValid on a `SymbolicStaticError`? How can this even happen?

**Comment by vladyslav-dubina** [comment]
In the tests above, we get here https://github.com/vladyslav-dubina/p4c/blob/3d5cd9d0ec8cfef5b3881503d35e33896723c5e9/midend/interpreter.cpp#L326 and a SymbolicStaticError is returned. And then we get here https://github.com/vladyslav-dubina/p4c/blob/3d5cd9d0ec8cfef5b3881503d35e33896723c5e9/midend/interpreter.cpp#L1102 , and in our case, base is equal to SymbolicStaticError

**Comment by fruffy** [comment]
This does not seem to make any sense to me. This seems like a design flaw that should be fixed instead of applying band-aid around the `MethodCallExpression` function. But I do not understand the design of this interpreter well enough. Here is my interpretation. It is legitimate to read a header from an invalid header union. In fact, 

```p4 
U u;
u.h1.setValid();     // u and u.h1 are both valid
H1 my_h1 = u.h1;     // my_h1 is now valid, but contains an undefined value
```
should be a common use case, right? Here you would return a `SymbolicStaticError`, which is incorrect. Instead, an uninitialized header is returned, which is then initialized using `setValid`. This will set both the union and the header to valid. How will this be done in this interpreter design?

**Comment by vladyslav-dubina** [comment]
If you need to return an uninitialized header, then you need to remove the check here https://github.com/vladyslav-dubina/p4c/blob/3d5cd9d0ec8cfef5b3881503d35e33896723c5e9/midend/interpreter.cpp#L326 ?

**Comment by fruffy** [comment]
Yes, it does not look right to me.

**Comment by vladyslav-dubina** [comment]
If I understood everything correctly, then we must do something like this.

**Comment by fruffy** [comment]
I am not sure if this is enough. You want to return a header that is uninitialized. Does this actually do that? Also, when the header is set to valid, is the union also valid? How does this work?

**Comment by vladyslav-dubina** [comment]
I need to think more about this question, probably more checks are needed in order to make sure if the union is initialized

**Comment by vladyslav-dubina** [comment]
Here "https://github.com/vladyslav-dubina/p4c/blob/38bf66b9cb9b7e6c91999aaac3f7f8e5e2669b74/midend/interpreter.cpp#L1105" we return an error, why can't we do the same thing, with the return of an error?

**Comment by fruffy** [comment]
After thinking about it for a while I think your original solution might be right. Considering this code https://github.com/vladyslav-dubina/p4c/blob/38bf66b9cb9b7e6c91999aaac3f7f8e5e2669b74/midend/interpreter.cpp#L420
So we really just need checks for symbolic errors, although I do not understand the implications here. 
@mbudiu-vmw What is the purpose of symbolic errors in this interpreter? Is a read from an invalid header or union a static error?

**Comment by mbudiu-vmw** [comment]
I wrote this code more than 5 years ago. All recent changes were by @VolodymyrPeschanenkoIntel. Ideally he should review these contributions.

**Comment by VolodymyrPeschanenkoIntel** [comment]
@fruffy , @mbudiu-vmw , We used interpreter.h for interpretation of the expression for parser unroll. For this case if we try to calculate value from invalid header the corresponded error should be generated. But in general interpretation of the setValid and setInvalid should be fixed, because in any case of valid/invalid of a headers the expression like "u.h1.setValid()" should be interpreted in the correct way. I have communicated with @vladyslav-dubina to find appreciate solution for this. Let's wait for the new implementation.

File: midend/interpreter.cpp [hunk 1]
```diff
@@ -1091,26 +1126,54 @@ void ExpressionEvaluator::postorder(const IR::MethodCallExpression* expression)
...
+          } else if (auto expr = expression->method->to<IR::Member>()->expr) {
+            node = expr;
+          }
+          auto structVar = get(node);
+          if (structVar->is<SymbolicHeader>()) {
+            auto header = structVar->to<SymbolicHeader>();
+            header->setValid(name == IR::Type_Header::setValid);
             set(expression, SymbolicVoid::get());
             return;
+          } else if (structVar->is<SymbolicHeaderUnion>()) {
```
[the remark this answers is not in the document]
Comment (vladyslav-dubina) [reply midend/interpreter.cpp:1149]: If I understand correctly, you need to check for valid `Header_Union ` elements in `setValid ` and set valid against this?

File: midend/interpreter.cpp [hunk 2]
```diff
@@ -1091,26 +1126,54 @@ void ExpressionEvaluator::postorder(const IR::MethodCallExpression* expression)
...
+          const IR::Expression *node;
+          cstring memberName;
+          bool flag = true;
+          if (auto member = expression->method->to<IR::Member>()
+                                ->expr->to<IR::Member>()) {
+            flag = false;
+            node = member->expr;
+            memberName = member->member.name;
+          } else if (auto expr = expression->method->to<IR::Member>()->expr) {
+            node = expr;
```
[the remark this answers is not in the document]
Comment (vladyslav-dubina) [reply midend/interpreter.cpp:1141]: I don't quite understand what exactly is not indented

[the remark this answers is not in the document]
Comment (fruffy) [reply midend/interpreter.cpp:1141]: This comment is outdated, indentation looks correct to me now.

File: midend/interpreter.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (vladyslav-dubina) [reply midend/interpreter.cpp:1152]: I made edits

**Comment by vladyslav-dubina** [comment]
@mbudiu-vmw  We have refactored the code and now PR ready for review

**Pull request closed without merging**